### PR TITLE
SRIOV for VMSets, without perf

### PR DESCRIPTION
**Pull request by abays**
Basic SRIOV integration for `OpenStackVmSet` and `OpenStackNet` CRDs.  Leaving performance operator integration and `OpenStackControlPlane` changes for other PRs in order to keep this one smaller.  With what is provided here, you can test SRIOV with VMSets against worker nodes allocated by the "virtual SRIOV" functionality in our dev-tools.

### CODE REVIEW

File: controllers/openstacknet_controller.go [hunk 1]
```diff
@@ -93,6 +95,48 @@ func (r *OpenStackNetReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error
...
+		return ctrl.Result{}, nil
+	}
+
+	if instance.Spec.AttachConfiguration.NodeSriovConfigurationPolicy.DesiredState.Port != "" {
+		if err := r.ensureSriov(instance); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
 	// generate NodeNetworkConfigurationPolicy
```
Comment (stuggi) [inline controllers/openstacknet_controller.go:140]: we probably need a check here to skip non sriov network configuration if there is no instance.Spec.AttachConfiguration.NodeNetworkConfigurationPolicy ?

Comment (abays) [reply controllers/openstacknet_controller.go:140]: Yep, forgot the `else` :)

File: config/rbac/role.yaml [hunk 2]
```diff
@@ -386,6 +386,19 @@ rules:
   - get
   - patch
   - update
+- apiGroups:
```
Comment (dprince) [inline config/rbac/role.yaml:389]: Once we land this the SRIOV operator is required right? Perhaps update update README with this requirement?

File: controllers/openstackvmset_controller.go [hunk 3]
```diff
@@ -91,6 +92,9 @@ func (r *OpenStackVMSetReconciler) GetScheme() *runtime.Scheme {
 // +kubebuilder:rbac:groups=osp-director.openstack.org,resources=openstackipsets,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=osp-director.openstack.org,resources=openstackipsets/finalizers,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=osp-director.openstack.org,resources=openstackipsets/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=osp-director.openstack.org,resources=openstacknets,verbs=get;list
+// FIXME: Is there a way to scope the following RBAC annotation to just the "openshift-sriov-network-operator" namespace?
```
Comment (dprince) [inline controllers/openstackvmset_controller.go:96]: The upstream docs do mention a 'namespace' option we could try? https://book.kubebuilder.io/reference/markers/rbac.html (perhaps this is outdated)

File: controllers/openstacknet_controller.go [hunk 4]
```diff
@@ -93,69 +98,287 @@ func (r *OpenStackNetReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error
...
-		return ctrl.Result{}, err
+
+	// If we have an SRIOV definition on this network, use that.  Otherwise assume the
+	// non-SRIOV configuration must be present
+	if instance.Spec.AttachConfiguration.NodeSriovConfigurationPolicy.DesiredState.Port != "" {
+		if err := r.ensureSriov(instance); err != nil {
+			return ctrl.Result{}, err
+		}
+	} else {
+		// generate NodeNetworkConfigurationPolicy
```
Comment (stuggi) [inline controllers/openstacknet_controller.go:144]: like with ensureSriov, we might want to move all the bellow to a method in a follow up patch

**Review by stuggi — APPROVED** [review]

**Pull request merged**